### PR TITLE
8329597: C2: Intrinsify Reference.clear

### PR DESCRIPTION
**Pull request by shipilev**
[JDK-8240696](https://bugs.openjdk.org/browse/JDK-8240696) added the native method for `Reference.clear`. The original patch skipped intrinsification of this method, because we thought `Reference.clear` is not on a performance sensitive path. However, it shows up prominently on simple benchmarks that touch e.g. `ThreadLocal` cleanups. See the bug for an example profile with `RRWL` benchmarks.

We need to know the actual oop strongness/weakness before we call into C2 Access API, this work models this after existing code for `refersTo0` intrinsics. C2 Access also need a support for `AS_NO_KEEPALIVE` for stores. 

Additional testing:
 - [x] Linux x86_64 server fastdebug, `all`
 - [x] Linux AArch64 server fastdebug, `all`

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329597](https://bugs.openjdk.org/browse/JDK-8329597): C2: Intrinsify Reference.clear (**Enhancement** - P4)

### Reviewers
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) Review applies to [9f7ad7ab](https://git.openjdk.org/jdk/pull/20139/files/9f7ad7ab5fae6baff02f6350fe4a70657c26edd6)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) Review applies to [9f7ad7ab](https://git.openjdk.org/jdk/pull/20139/files/9f7ad7ab5fae6baff02f6350fe4a70657c26edd6)
 * [Roberto Castañeda Lozano](https://openjdk.org/census#rcastanedalo) (@robcasloz - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20139/head:pull/20139` \
`$ git checkout pull/20139`

Update a local copy of the PR: \
`$ git checkout pull/20139` \
`$ git pull https://git.openjdk.org/jdk.git pull/20139/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20139`

View PR using the GUI difftool: \
`$ git pr show -t 20139`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20139.diff">https://git.openjdk.org/jdk/pull/20139.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20139#issuecomment-2223708798)

### CODE REVIEW

**Comment by shipilev** [comment]
On Mac AArch64, which experiences heavy WX transition:

```

```

On x86_64 m7g.16xlarge, which only suffers from the native call:

```
# Intrinsic OFF
ReferenceClear.phantom      avgt    9  14.643 ± 0.049  ns/op
ReferenceClear.soft         avgt    9  14.939 ± 0.438  ns/op
ReferenceClear.weak         avgt    9  14.648 ± 0.081  ns/op
ReferenceClear.phantom_new  avgt    9  19.859 ± 2.405  ns/op
ReferenceClear.soft_new     avgt    9  20.208 ± 1.805  ns/op
ReferenceClear.weak_new     avgt    9  20.385 ± 2.570  ns/op

# Intrinsic ON (default)
ReferenceClear.phantom      avgt    9  0.821 ± 0.010  ns/op
ReferenceClear.soft         avgt    9  0.817 ± 0.007  ns/op
ReferenceClear.weak         avgt    9  0.819 ± 0.010  ns/op
ReferenceClear.phantom_new  avgt    9  4.195 ± 0.729  ns/op
ReferenceClear.soft_new     avgt    9  4.315 ± 0.599  ns/op
ReferenceClear.weak_new     avgt    9  3.986 ± 0.596  ns/op
```

**Comment by fisk** [comment]
The reason we did not do this before is that this is not a strong reference store. Strong reference stores with a SATB collector will keep the referent alive, which is typically the exact opposite of what a user wants when they clear a Reference.

**Comment by shipilev** [comment]
> The reason we did not do this before is that this is not a strong reference store. Strong reference stores with a SATB collector will keep the referent alive, which is typically the exact opposite of what a user wants when they clear a Reference.

You mean not doing this store just on the Java side? Yes, I agree, it would be awkward. In intrinsic, we are storing with the same decorators that `JVM_ReferenceClear` is using, which should be good with SATB collectors. Perhaps I am misunderstanding the comment.

**Comment by fisk** [comment]
> > The reason we did not do this before is that this is not a strong reference store. Strong reference stores with a SATB collector will keep the referent alive, which is typically the exact opposite of what a user wants when they clear a Reference.
> 
> You mean not doing this store just on the Java side? Yes, I agree, it would be awkward. In intrinsic, we are storing with the same decorators that `JVM_ReferenceClear` is using, which should be good with SATB collectors. Perhaps I am misunderstanding the comment.

The runtime use of the Access API knows how to resolve an unknown oop ref strength using AccessBarrierSupport::resolve_unknown_oop_ref_strength. However, we do not have support for that in the C2 backend. In fact, it does not understand non-strong oop stores at all. Because there hasn't really been a use case for it, other than clearing a Reference. That's the precise reason why we do not have a clear intrinsic; it would have to add that infrastructure.

**Comment by shipilev** [comment]
> The runtime use of the Access API knows how to resolve an unknown oop ref strength using AccessBarrierSupport::resolve_unknown_oop_ref_strength. However, we do not have support for that in the C2 backend. In fact, it does not understand non-strong oop stores at all. 

Aw, nice usability landmine. I thought C2 barrier set would assert on me if it cannot deliver. Apparently not, I see it just does pre-barriers when it is not sure what strongness the store is. Hrmpf. OK, let me see what can be done here.

**Comment by kimbarrett** [comment]
> > The runtime use of the Access API knows how to resolve an unknown oop ref strength using AccessBarrierSupport::resolve_unknown_oop_ref_strength. However, we do not have support for that in the C2 backend. In fact, it does not understand non-strong oop stores at all.
> 
> Aw, nice usability landmine. I thought C2 barrier set would assert on me if it cannot deliver. Apparently not, [...]

Reference.refersTo has similar issues.  See refersToImpl and refersTo0 in both Reference and PhantomReference.
I think you should be able to model on those and the intrinsic implementation for refersTo to get what you want.

**Comment by kimbarrett** [comment]
> > Aw, nice usability landmine. I thought C2 barrier set would assert on me if it cannot deliver. Apparently not, [...]
> 
> Reference.refersTo has similar issues. See refersToImpl and refersTo0 in both Reference and PhantomReference. I think you should be able to model on those and the intrinsic implementation for refersTo to get what you want.
> 
> One additional complication is that Reference.enqueue intentionally calls clear0. If implementing clear similarly to refersTo, then enqueue should be changed to call clearImpl.

I should have read what I was replying to more carefully, rather than focusing on what was further up in the thread.
Looks like you (@shipilev) already spotted the refersTo stuff.  But the enqueue => clear0 could have easily been missed,
so perhaps not an entirely unneeded suggestion.

**Comment by shipilev** [comment]
> I should have read what I was replying to more carefully, rather than focusing on what was further up in the thread. Looks like you (@shipilev) already spotted the refersTo stuff. But the enqueue => clear0 could have easily been missed, so perhaps not an entirely unneeded suggestion.

Yeah, thanks. The `enqueue => clear0` was indeed easy to miss.

Pushed the crude prototype that follows `refersTo` example and drills some new `AS_NO_KEEPALIVE` holes in C2 Access API to cover this intrinsic case. Super untested. IR tests are still failing, I'll take more in-depth look there. (Perhaps it would not be possible to clearly match the absence of pre-barrier in IR tests, we'll see.)

**Comment by shipilev** [comment]
Split out the `refersTo` test to https://github.com/openjdk/jdk/pull/20215.

**Comment by shipilev** [comment]
Yeah, so this version seems to work well on tests. I am being extra paranoid about only accepting `null` stores, since `AS_NO_KEEPALIVE` means all other barriers like inter-generational post-barriers in G1 should still work.

**Comment by shipilev** [comment]
Not now, bot. Still waiting for reviews.

**Comment by vnkozlov** [comment]
Is ZGC affected by this? I see only G1 and Shenandoah changes.

**Comment by shipilev** [comment]
> Is ZGC affected by this? I see only G1 and Shenandoah changes.

Good question.

ZGC expands the GC barriers late. This is why the IR test configuration that tests ZGC shows the same result as with other collectors: no additional fluff in IR. I would not expect we need anything else in late expansion for ZGC for Reference.clear, but maybe I am tired and cannot see it. Can you confirm this is fine, @fisk?

**Comment by vnkozlov** [comment]
There is coming JEP for later G1 barriers expansion similar to ZGC. Will you still need this after it? I assume Shenandoah will follow G1 later.

File: src/java.base/share/classes/java/lang/ref/Reference.java
Comment (kimbarrett) [inline src/java.base/share/classes/java/lang/ref/Reference.java:420]: Description of clear0 is rendered stale by this change.  The first sentence is no longer true, since it's now
clearImpl that has that role.  The second sentence probably ought to also be moved into the description of
clearImpl.

**Review by kimbarrett — CHANGES_REQUESTED** [review]

**Comment by fisk** [comment]
> > Is ZGC affected by this? I see only G1 and Shenandoah changes.
> 
> Good question.
> 
> ZGC expands the GC barriers late. This is why the IR test configuration that tests ZGC shows the same result as with other collectors: no additional fluff in IR. I would not expect we need anything else in late expansion for ZGC for Reference.clear, but maybe I am tired and cannot see it. Can you confirm this is fine, @fisk?

ZGC needs some changes. Without doing anything, we propagate the AS_NO_KEEPALIVE decorator to the corresponding ZBarrierNoKeepalive bit being set in the barrier data of the StorePNode. However, we don't really do anything special with that information and we will in practice end up keeping the referent alive when clearing it with generational ZGC. The point of introducing the native implementation in the first place, was to make sure our GCs don't keep the referent alive when clearing it, as the user intention is clearly to not keep it alive.

I think we need a new ZBarrierSetRuntime::no_keepalive_store_barrier_on_oop_field_without_healing(oop* p) and to make that the selected slow path function when ZBarrierNoKeepalive is set on a StorePNode. Its implementation would call ZBarrier::no_keep_alive_store_barrier_on_heap_oop_field. This should do the trick.

Please let me know if you need further assistance.

**Comment by shipilev** [comment]
> I think we need a new ZBarrierSetRuntime::no_keepalive_store_barrier_on_oop_field_without_healing(oop* p) and to make that the selected slow path function when ZBarrierNoKeepalive is set on a StorePNode. Its implementation would call ZBarrier::no_keep_alive_store_barrier_on_heap_oop_field. This should do the trick.

Thanks! See new commits: is that the shape you were thinking of? Once we get AArch64 parts right, I'll copy-paste that to other arches.

File: src/java.base/share/classes/java/lang/ref/Reference.java
Comment (shipilev) [reply src/java.base/share/classes/java/lang/ref/Reference.java]: Thanks! I tightened up comments a bit, take another look?

Comment (kimbarrett) [reply src/java.base/share/classes/java/lang/ref/Reference.java]: Yes, that's better.  Thanks.

**Review by kimbarrett — COMMENTED** [review]
Removing my "Request changes" as request has been satisfied.

I've only really looked at the changes in java.base, which look fine.  I've skimmed some
of the compiler code, but don't feel qualified to properly review it.  So don't count or
wait for me as a reviewer.

File: test/micro/org/openjdk/bench/java/lang/ref/ReferenceClear.java
Comment (TobiHartmann) [inline test/micro/org/openjdk/bench/java/lang/ref/ReferenceClear.java:2]: Drive-by comment: The `// *` format looks weird.

File: src/hotspot/share/opto/library_call.cpp
Comment (fisk) [inline src/hotspot/share/opto/library_call.cpp:7002]: I think this CPU memory barrier and comment above are confusing and obviously just taken from the referent loading intrinsics. The commoning it is talking about is to short circuit a second load with the result of a first load of the referent field, since the compiler "knows" the first load would give the "same" answer unless "something happened" (GC clearing it).
In this case the mutator just cleared it, so what the compiler thinks is that null is stored in that field. And that's completely accurate, and the GC is not going to transition the field from null to some random other object.
Let's remove this CPU memory barrier and the misleading comments.

**Review by fisk — CHANGES_REQUESTED** [review]

File: test/micro/org/openjdk/bench/java/lang/ref/ReferenceClear.java
Comment (shipilev) [reply test/micro/org/openjdk/bench/java/lang/ref/ReferenceClear.java:2]: Actually, this constellation of single-line comments should be replaced with a multi-line comment block. The fix is now in my (unpublished) patch queue, thanks.

File: src/hotspot/share/opto/library_call.cpp
Comment (shipilev) [reply src/hotspot/share/opto/library_call.cpp:7002]: Right. I removed it in my local patch queue. Now I have to reconcile this PR with late barrier expansion in G1, I'll push the update once that is done.

File: test/micro/org/openjdk/bench/java/lang/ref/ReferenceClear.java
Comment (TobiHartmann) [reply test/micro/org/openjdk/bench/java/lang/ref/ReferenceClear.java]: Yes, that's what I meant. Thanks.

**Comment by shipilev** [comment]
Tests pass with the new change. I eyeballed G1 perfasm output on new benchmark, and there are no barriers in sight as well.

File: src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp
Comment (shipilev) [reply src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp:342]: Right. We also do not need this complexity in Shenandoah barriers. This check was dragged here from the load barriers that _want_ to check if we are reading the `Reference.referent` and feed it to SATB _unless_ there is a no-keep-alive. For store barriers it is unnecessary, and we can just do keep-alive checks straight up. Should be done in new commit, testing now.

**Review by fisk — APPROVED** [review]
Looks good.

**Comment by shipilev** [comment]
Thanks! I see Kim reviewed JDK parts, so we need another Reviewer for Hotspot parts.

**Review by vnkozlov — APPROVED** [review]
C2 change (intrinsics code) is fine.

**Comment by robcasloz** [comment]
> Thanks for review, folks. I am re-running testing locally here. Would appreciate if you can give this patch a spin through your CIs as well.

I will run some internal CI testing and report back in one or two days.

File: src/hotspot/share/gc/z/zBarrierSetRuntime.hpp
Comment (mur47x111) [inline src/hotspot/share/gc/z/zBarrierSetRuntime.hpp:43]: Could you please export this to JVMCI? I.e.,
```
diff --git a/src/hotspot/share/jvmci/vmStructs_jvmci.cpp b/src/hotspot/share/jvmci/vmStructs_jvmci.cpp
index 5452cca96b8..46aeb996c56 100644
--- a/src/hotspot/share/jvmci/vmStructs_jvmci.cpp
+++ b/src/hotspot/share/jvmci/vmStructs_jvmci.cpp
@@ -847,6 +847,7 @@
   ZGC_ONLY(DECLARE_FUNCTION_FROM_ADDR(declare_function_with_value, ZBarrierSetRuntime::load_barrier_on_oop_field_preloaded_store_good))           \
   ZGC_ONLY(DECLARE_FUNCTION_FROM_ADDR(declare_function_with_value, ZBarrierSetRuntime::no_keepalive_load_barrier_on_weak_oop_field_preloaded))    \
   ZGC_ONLY(DECLARE_FUNCTION_FROM_ADDR(declare_function_with_value, ZBarrierSetRuntime::no_keepalive_load_barrier_on_phantom_oop_field_preloaded)) \
+  ZGC_ONLY(DECLARE_FUNCTION_FROM_ADDR(declare_function_with_value, ZBarrierSetRuntime::no_keepalive_store_barrier_on_oop_field_without_healing)   \
   ZGC_ONLY(DECLARE_FUNCTION_FROM_ADDR(declare_function_with_value, ZBarrierSetRuntime::store_barrier_on_native_oop_field_without_healing))        \
   ZGC_ONLY(DECLARE_FUNCTION_FROM_ADDR(declare_function_with_value, ZBarrierSetRuntime::store_barrier_on_oop_field_with_healing))                  \
   ZGC_ONLY(DECLARE_FUNCTION_FROM_ADDR(declare_function_with_value, ZBarrierSetRuntime::store_barrier_on_oop_field_without_healing))               \
```
Thanks!

Comment (shipilev) [reply src/hotspot/share/gc/z/zBarrierSetRuntime.hpp:43]: Done!

**Comment by robcasloz** [comment]
> I will run some internal CI testing and report back in one or two days.

The test results look good. I tested the changes (up to commit 9f7ad7ab) on top of jdk-24+19 running tier1-tier5 on all Oracle-supported platforms.

**Review by robcasloz — APPROVED** [review]

**Comment by shipilev** [comment]
Thank you for testing! Here goes.

/integrate

**Pull request closed without merging**
